### PR TITLE
recent projects: cleanup ui

### PR DESCRIPTION
**Pull request by bennetbo**
As the ui for the file finder was recently changed in #7364, I think it makes sense to also update the ui of the recent projects overlay.

Before:
image

After:
image

Release Notes:

- Improved UI of recent project overlay

### CODE REVIEW

File: crates/recent_projects/src/recent_projects.rs [hunk 1]
```diff
@@ -235,13 +235,16 @@ impl PickerDelegate for RecentProjectsDelegate {
...
-                    v_flex()
-                        .child(highlighted_location.names)
-                        .when(self.render_paths, |this| {
-                            this.children(highlighted_location.paths)
-                        }),
-                ),
+                .child(h_flex().gap_2().child(highlighted_location.names).when(
+                    self.render_paths,
+                    |this| {
+                        this.children(highlighted_location.paths.into_iter().map(|path| {
```
Comment (bennetbo) [inline crates/recent_projects/src/recent_projects.rs:241]: Can there actually be multiple paths for an entry? I didn't find a case where that is happening. If it can actually happen this might look bad, because the paths are laid out horizontally.

File: crates/recent_projects/src/recent_projects.rs [hunk 2]
```diff
@@ -139,7 +139,7 @@ impl PickerDelegate for RecentProjectsDelegate {
     type ListItem = ListItem;
 
     fn placeholder_text(&self) -> Arc<str> {
-        "Recent Projects...".into()
+        "Search recent projects...".into()
```
Comment (bennetbo) [inline crates/recent_projects/src/recent_projects.rs:142]: I changed this because it is more in line with the rest of the ui. "Search..." is used in the file finder, channel manage members overlay etc.

File: crates/recent_projects/src/recent_projects.rs (hunk 1 above)
Comment (bennetbo) [reply crates/recent_projects/src/recent_projects.rs:241]: Aha thanks, I thought I was missing something there. I never used the "add folder to workspace" feature. I will try to make some improvements then.

Comment (mikayla-maki) [reply crates/recent_projects/src/recent_projects.rs:241]: Yeah, though the multiple paths only show up if your most recent path (the top one) has multiple paths.

**Comment by bennetbo** [comment]
Thanks for the comments, I adjusted the layout to look nice for multiple paths

Before:
image

After:
image

**Review by SomeoneToIgnore — COMMENTED** [review]
I have a beautiful font (Monocraft) that does horrendous things to Zed's layouts.
So, I quickly checked a few things with this branch and posting it here for visibility.
Original Zed is not much better, but my issue with this PR is that it changes those things so that they are broken different way now, without fixing them


* Long file paths are trimmed (see charts-repo-actions-demo entry):
image

Original Zed is much better here since the path below has enough space:
Screenshot 2024-02-08 at 10 13 58

* Multi dir paths force single dir entries to get too tall and all other obvious layouting issues: 

image

Original Zed is much worse but it's not like the new version is qualitatively better:

Screenshot 2024-02-08 at 10 13 48

also, as it's shown on the other original Zed screenshot above, it is able to show multidir project entry normally, if the picker selection changes. So I think original Zed is somewhat closer to being normal than the new version.

**Comment by mikayla-maki** [comment]
> but it's not like the new version is qualitatively better

I would disagree, this version at least shows all of the data without being completely broken.

**Comment by SomeoneToIgnore** [comment]
I would disagree: the new version does not show "all the data", trimming the directories — that is my issue with both versions.

**Comment by SomeoneToIgnore** [comment]
image

I've checked the new version, and it looks almost done to me: there are still long paths that are trimmed, but the new hover thing makes it much better.

One last thing that seems to be broken is the list scroll, but since the PR is a draft, you might be aware of this 🙂

**Comment by bennetbo** [comment]
Thanks @SomeoneToIgnore for testing. Here is a summary of the implementation so far:
- Switched back to the vertical layout, because longer paths are easier to display
- Paths get now displayed in muted color
- There is now a new variant of the `Picker` which allows items to have different sizes (`Picker::list(...)`)
image
- Added tooltip to show the paths of each project, so when they are cut off you can still see where they point to. Which has the nice side effect that you can actually see the paths of a project when using the project switch button in the upper left corner:
image


@SomeoneToIgnore yes im aware of the scrolling issue and that's what im currently struggling with.

The list should take all the available space inside the overlay, up to a maximum of what's passed into picker. However if i set the `list`'s size to `size_full()` it doesn't take any space (just 0). This works fine when using `uniform_list`.

 The only way I can force the list to display is to set a fixed width.
If anyone would be able to help me with this that would be greatly appreciated. 
Also thanks to @RemcoSmitsDev for investigating the list implementation with me and trying to solve the layout issue.


Some other ideas which might be nice to have:
- Remove duplicated project name from path (`zed    ~/workspace` instead of `zed  ~/workspace/zed` )
- Instead of cutting of the characters of the path/project name display some kind of `...` (I don't think this is possible in gpui so far though, so probably out of scope)

**Comment by SomeoneToIgnore** [comment]
> the `list`'s size to `size_full()` it doesn't take any space (just 0)

I wonder if it's the same as https://github.com/zed-industries/zed/pull/6958#discussion_r1469511500 , where we need to wrap a list with canvas and set their size to full.
But alas, I'm not good with gpui layouts so cannot help you here much.

**Comment by SomeoneToIgnore** [comment]
Oh, sorry for putting you through all this, gpui can hurt sometimes.
I still see something odd with the picker from time to time — if I toggle it, sometimes it hangs in a "collapsed" state for a few seconds, but then gets back to normal if I interact with it or wait a bit:

image

I think that overall, this is still an improvement, but fear that this bug is more noticeable than before.
On the other hand, I've spent some time nitpicking here already, so not sure how much should I continue with that.
@ConradIrwin do you think we can merge this and fix the issue later? Or maybe you have an idea how to fix that last bit?

**Comment by bennetbo** [comment]
Yes, those are the edge cases that im talking about:
- When typing something into the search there seems to be some portion of the list that is still getting rendered  (this is the one you screenshotted)
- When the list takes less size then the max height the bottom padding seems to be to large (Looks like its doubled)
Screenshot 2024-02-14 at 21 54 41
- It feels like its less responsive (this might be because we need to paint once to report the correct size, so we are missing a frame)

I agree with you, that flickering is noticeable and I feel like we do not want to merge this until those pain points are resolved!

**Comment by bennetbo** [comment]
> When typing something into the search there seems to be some portion of the list that is still getting rendered (this is the one you screenshotted)

I fixed this, although the visuals are not quite the same as for uniform list. Uniform list will not show up at all, while List will still be rendered (only with the padding) but not show any items

> When the list takes less size then the max height the bottom padding seems to be to large (Looks like its doubled)

I took me way too long to figure out that there is no need to include the padding size in the `request_layout` function (which makes sense)

I believe the visual "bugs" are gone, however the 1 or 2 missing frames while typing really bother me (zed is always so fast and responsive, this doesn't fit in). I will look into it tomorrow.

**Comment by SomeoneToIgnore** [comment]
Thank you for keeping squashing them all, that's the hardest part.

I can only confirm the occasional blank panel state (as mentioned above) and an offset issue when scrolling up:
Initial (good): 
image
After pressing up (meh): 
image

**Comment by bennetbo** [comment]
Thanks for the testing once again. This was happening because there was already a max_height on the picker itself, which I used as a workaround for specifying the max_height for the list as well (which is a little too small, because the search panel takes some space). Therefore the scroll area "thinks" its larger then it actually is -> no scrolling.

By reworking the layout implementation I was able to fix the flickering issues, but im afraid I am now stuck on another problem:

The new list implementation fixes the flickering by measuring items in the layout phase (this is cached so there is not much more overhead than before).
I tried to keep the same approach for layout that `uniform_list` is using. However I don't get the correct results currently...

Here is the layout "trace" for `uniform_list` and `list`, each invocation of the layout callback prints the `known_dimensions`, `available_space`, `width` and `height`. As you can see the `uniform_list` callback is called more often then the `list` one and is missing  crucial information about the max_height of the parent. Take a look at the following trace:

```
// UNIFORM LIST

Layout
[crates/gpui/src/elements/uniform_list.rs:128] known_dimensions = Size { None × None }
[crates/gpui/src/elements/uniform_list.rs:128] available_space = Size { MinContent × MinContent }
[crates/gpui/src/elements/uniform_list.rs:146] width = 220 px
[crates/gpui/src/elements/uniform_list.rs:146] height = 350 px
[crates/gpui/src/elements/uniform_list.rs:128] known_dimensions = Size { None × Some(288 px) }
[crates/gpui/src/elements/uniform_list.rs:128] available_space = Size { MinContent × Definite(272 px) }
[crates/gpui/src/elements/uniform_list.rs:146] width = 220 px
[crates/gpui/src/elements/uniform_list.rs:146] height = 272 px
[crates/gpui/src/elements/uniform_list.rs:128] known_dimensions = Size { Some(542 px) × None }
[crates/gpui/src/elements/uniform_list.rs:128] available_space = Size { Definite(542 px) × MinContent }
[crates/gpui/src/elements/uniform_list.rs:146] width = 542 px
[crates/gpui/src/elements/uniform_list.rs:146] height = 350 px
[crates/gpui/src/elements/uniform_list.rs:128] known_dimensions = Size { None × None }
[crates/gpui/src/elements/uniform_list.rs:128] available_space = Size { Definite(542 px) × MinContent }
[crates/gpui/src/elements/uniform_list.rs:146] width = 542 px
[crates/gpui/src/elements/uniform_list.rs:146] height = 350 px
[crates/gpui/src/elements/uniform_list.rs:128] known_dimensions = Size { None × Some(288 px) }
[crates/gpui/src/elements/uniform_list.rs:128] available_space = Size { Definite(542 px) × Definite(272 px) }
[crates/gpui/src/elements/uniform_list.rs:146] width = 542 px
[crates/gpui/src/elements/uniform_list.rs:146] height = 272 px
Paint
[crates/gpui/src/elements/uniform_list.rs:167] bounds = Bounds {
    origin: Point {
        x: 369 px,
        y: 152 px,
    },
    size: Size { 542 px × 288 px },
}

// LIST

[crates/gpui/src/elements/list.rs:516] known_dimensions = Size { None × None }
[crates/gpui/src/elements/list.rs:516] available_space = Size { MinContent × MinContent }
[crates/gpui/src/elements/list.rs:529] width = 788 px
[crates/gpui/src/elements/list.rs:529] height = 447 px
[crates/gpui/src/elements/list.rs:516] known_dimensions = Size { Some(542 px) × None }
[crates/gpui/src/elements/list.rs:516] available_space = Size { Definite(542 px) × MinContent }
[crates/gpui/src/elements/list.rs:529] width = 542 px
[crates/gpui/src/elements/list.rs:529] height = 447 px
[crates/gpui/src/elements/list.rs:516] known_dimensions = Size { None × Some(463 px) }
[crates/gpui/src/elements/list.rs:516] available_space = Size { Definite(542 px) × Definite(447 px) }
[crates/gpui/src/elements/list.rs:529] width = 542 px
[crates/gpui/src/elements/list.rs:529] height = 447 px
Paint
[crates/gpui/src/elements/list.rs:547] bounds = Bounds {
    origin: Point {
        x: 369 px,
        y: 152 px,
    },
    size: Size { 542 px × 463 px },
}
```

My current understanding:
When reporting a height that is larger in the max height of the container, the `request_measured_layout` callback should be called again with `known_dimensions.height` set to the maximum height of the parent container (`288px/272 px without padding` in this case).
This is what's happening for `uniform_list`, but although I pretty much implemented the same for `list`, taffy never seems to call the `request_measured_layout` callback with:
[crates/gpui/src/elements/uniform_list.rs:128] known_dimensions = Size { None × Some(288 px) }
[crates/gpui/src/elements/uniform_list.rs:128] available_space = Size { MinContent × Definite(272 px) }
which is the information we need to be able to shrink down the list.

I hope my explanation makes sense, im not sure what's the problem here.
It might just be that im not that familiar with the gpui code, so it would be awesome if anyone could jump in and help. Also thanks for bearing with me while working on this.

**Comment by SomeoneToIgnore** [comment]
Well this seems to be again very similar to https://github.com/zed-industries/zed/pull/6958#discussion_r1469511500 comment: at least with my "not scrolling" issue reported there, there was a very similar taffy behavior that did not have any max_height dimensions set.
If that `canvas` approach does not help, I'm afraid that I'm pretty clueless and can only propose to ask another maintainer to help here, as my gpui layouting skills are also very minimal.

**Comment by bennetbo** [comment]
I managed to finally figure it out! Apparently there is a `overflow` property on style which I wasn't setting correctly. 
I believe it is finally working, I need to do some testing.

**Comment by bennetbo** [comment]
My changes seem to break existing layouts using `list`, will look into it tomorrow.

**Comment by bennetbo** [comment]
Should be ready to review now, you can see a showcase of the current state below:
https://github.com/zed-industries/zed/assets/53836821/c37585eb-9936-4dd6-95b2-5e9e391e9670

Summary:
- Handle items with different sizes in the picker correctly
- Add an alternative approach for calculating `list` size based on its children (which we need for auto height)
- Show tooltip for paths, so you can still see them when they are cut off
- Improved the ui (e.g. paths are now displayed in muted color)

A few notes on the implementation:
Im not sure about the design off the list layout code. I introduced an alternative codepath for calculating the size of the list based on its children (`ListSizingBehavior::Infer`). This was the only solution I was able to come up with to support the new behaviour and not break existing layouts. Would be great to get feedback on this.

**Review by SomeoneToIgnore — COMMENTED** [review]
I believe your approach is what it has to be, after all, so I'm fine with a separate codepath being used.

The recent file picker looks really polished now, thank you for coming up with all those.
Let's fix one big degradation with the rest of the pickers and we'll merge this.

File: crates/command_palette/src/command_palette.rs [hunk 3]
```diff
@@ -80,7 +80,7 @@ impl CommandPalette {
             previous_focus_handle,
         );
 
-        let picker = cx.new_view(|cx| Picker::new(delegate, cx));
+        let picker = cx.new_view(|cx| Picker::uniform_list(delegate, cx));
```
[the remark this answers is not in the document]
Comment (bennetbo) [reply crates/command_palette/src/command_palette.rs:83]: Fixed this in the latest commit. The uniform list was not taking padding into account when scrolling to a specific item. 

I would advise against using `Picker::list` everywhere, as the uniform_list implementation is more performant. uniform_list can optimise the layout process (if the assumption that all items are the same height is true), by only measuring a single item and then multiplying the height by that. list can't do that, because items have different heights, therefore it needs to call measure on pretty much every item that is in view (bounds + overdraw). 
So I suggest keeping `Picker::uniform_list` everywhere except the recent projects picker.

TLDR: We should only be using `Picker::list` when there are items with different heights. I believe that recent projects is the only place were this is true right now.

[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply crates/command_palette/src/command_palette.rs:83]: Looks like the `clippy` is unhappy now, so while we are at it, let's add a rustdoc with those considerations for both `Picker` constructor methods then?
It was not very obvious for me, so I assume more people will find the explanations useful.

[the remark this answers is not in the document]
Comment (bennetbo) [reply crates/command_palette/src/command_palette.rs:83]: What is `clippy` complaining about? Im not able to see any logs in the linux CI. I added some documentation to `Picker::list` and `Picker::uniform_list`.

[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply crates/command_palette/src/command_palette.rs:83]: Sorry, I did not look at the actual CI run, but seems that it failed due to out of disk space for the Linux task, so disregard that.

Thank you for the docs nonetheless.

[the remark this answers is not in the document]
Comment (bennetbo) [reply crates/command_palette/src/command_palette.rs:83]: All good, looks like it is passing now.

**Review by SomeoneToIgnore — APPROVED** [review]
Gave it another look and failed to find anything bad, nice! 🎉

Sorry for being a rather good filter but not a good source of help, at least now you seem to be pretty comfortable with gpui as a reward.

**Comment by bennetbo** [comment]
> Gave it another look and failed to find anything bad, nice! 🎉
> 
> Sorry for being a rather good filter but not a good source of help, at least now you seem to be pretty comfortable with gpui as a reward.

All good, thanks for all the testing you did, that was very helpful!
